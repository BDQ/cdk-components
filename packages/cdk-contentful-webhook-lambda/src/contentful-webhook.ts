import { CloudFormationCustomResourceEvent } from 'aws-lambda';
import * as contentful from 'contentful-management';

import {
    customResourceHelper,
    OnCreateHandler,
    OnUpdateHandler,
    OnDeleteHandler,
    ResourceHandler,
    ResourceHandlerReturn,
} from 'custom-resource-helper';

export interface WebhookProps {
    accessToken: string;
    spaceId: string;
    name: string;
    url: string;
    topics: string[];
}

const getProperties = (
    event: CloudFormationCustomResourceEvent,
): WebhookProps => {
    const props = event.ResourceProperties;

    const accessToken = props.AccessToken;
    const spaceId = props.SpaceId;
    const url = props.Url;
    const name = props.Name;
    const topics = props.Topics;

    return {
        accessToken,
        spaceId,
        url,
        name,
        topics,
    };
};

const getSpace = async (accessToken: string, spaceId: string) => {
    const client = contentful.createClient({
        accessToken,
    });

    return client.getSpace(spaceId);
};

const handleCreate: OnCreateHandler = async (
    event,
    _,
): Promise<ResourceHandlerReturn> => {
    const { accessToken, spaceId, ...props } = getProperties(event);

    const space = await getSpace(accessToken, spaceId);

    const responseData = await space.createWebhook({
        ...props,
    });

    const physicalResourceId = responseData.sys.id;

    return {
        physicalResourceId,
        responseData,
    };
};

const handleUpdate: OnUpdateHandler = async (
    event,
    _,
): Promise<ResourceHandlerReturn> => {
    const { accessToken, spaceId, ...props } = getProperties(event);

    const webhookId = event.PhysicalResourceId;

    const space = await getSpace(accessToken, spaceId);

    let webhook = await space.getWebhook(webhookId);

    webhook = {
        ...webhook,
        ...props,
    };

    const responseData = await webhook.update();

    const physicalResourceId = responseData.sys.id;

    return {
        physicalResourceId,
        responseData,
    };
};

const handleDelete: OnDeleteHandler = async (event, _): Promise<void> => {
    const { accessToken, spaceId } = getProperties(event);

    const webhookId = event.PhysicalResourceId;

    const space = await getSpace(accessToken, spaceId);

    const webhook = await space.getWebhook(webhookId);

    await webhook.delete();
};

export const handler = customResourceHelper(
    (): ResourceHandler => ({
        onCreate: handleCreate,
        onUpdate: handleUpdate,
        onDelete: handleDelete,
    }),
);
