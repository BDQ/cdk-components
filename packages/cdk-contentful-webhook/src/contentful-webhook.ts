import * as path from 'path';
import { Construct, Duration, CustomResource } from '@aws-cdk/core';
import { SingletonFunction, Runtime, Code } from '@aws-cdk/aws-lambda';

export interface ContentfulWebhookProps {
    accessToken: string;
    spaceId: string;
    name: string;
    url: string;
    topics: string[];
    logLevel?: 'debug' | 'info' | 'warning' | 'error';
}

export class ContentfulWebhook extends Construct {
    public constructor(
        scope: Construct,
        id: string,
        props: ContentfulWebhookProps,
    ) {
        super(scope, id);

        const handler = new SingletonFunction(this, 'CustomResourceHandler', {
            uuid: '91f2075f-b950-4743-a66b-ee0f6febf50d',
            runtime: Runtime.NODEJS_10_X,
            code: Code.fromAsset(
                path.join(__dirname, '..', 'lambda', 'bundle.zip'),
            ),
            handler: 'lib/contentful-webhook.handler',
            lambdaPurpose: 'Custom::ContentfulWebhook',
            timeout: Duration.minutes(15),
        });

        new CustomResource(this, 'CustomResource', {
            serviceToken: handler.functionArn,
            resourceType: 'Custom::ContentfulWebhook',
            pascalCaseProperties: true,
            properties: {
                ...props,
            },
        });
    }
}
