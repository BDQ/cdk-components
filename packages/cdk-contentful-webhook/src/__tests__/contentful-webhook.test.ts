import { Stack } from '@aws-cdk/core';
import 'jest-cdk-snapshot';

import { ContentfulWebhook } from '../contentful-webhook';

describe('cdk-contentful-webhook', (): void => {
    it('snapshot', (): void => {
        const stack = new Stack();

        new ContentfulWebhook(stack, 'ContentfulWebhook', {
            accessToken: '<access_token>',
            spaceId: '<space_id>',
            name: 'My webhook',
            url: 'https://www.example.com/test',
            topics: [
                'Entry.create',
                'ContentType.create',
                '*.publish',
                'Asset.*',
            ],
        });

        expect(stack).toMatchCdkSnapshot();
    });
});
