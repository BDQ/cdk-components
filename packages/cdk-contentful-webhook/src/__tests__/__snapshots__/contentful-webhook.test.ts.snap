// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdk-contentful-webhook snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters69f1a6347d9ec8846b5238785f2600d4d4498066a39df6942b0b4b7416efa262ArtifactHash8EA4FF52": Object {
      "Description": "Artifact hash for asset \\"69f1a6347d9ec8846b5238785f2600d4d4498066a39df6942b0b4b7416efa262\\"",
      "Type": "String",
    },
    "AssetParameters69f1a6347d9ec8846b5238785f2600d4d4498066a39df6942b0b4b7416efa262S3Bucket18224C47": Object {
      "Description": "S3 bucket for asset \\"69f1a6347d9ec8846b5238785f2600d4d4498066a39df6942b0b4b7416efa262\\"",
      "Type": "String",
    },
    "AssetParameters69f1a6347d9ec8846b5238785f2600d4d4498066a39df6942b0b4b7416efa262S3VersionKeyED23AA03": Object {
      "Description": "S3 key for asset version \\"69f1a6347d9ec8846b5238785f2600d4d4498066a39df6942b0b4b7416efa262\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ContentfulWebhookCustomResourceFC4E1F73": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AccessToken": "<access_token>",
        "Name": "My webhook",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomContentfulWebhook91f2075fb9504743a66bee0f6febf50d4005AAA1",
            "Arn",
          ],
        },
        "SpaceId": "<space_id>",
        "Topics": Array [
          "Entry.create",
          "ContentType.create",
          "*.publish",
          "Asset.*",
        ],
        "Url": "https://www.example.com/test",
      },
      "Type": "Custom::ContentfulWebhook",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomContentfulWebhook91f2075fb9504743a66bee0f6febf50d4005AAA1": Object {
      "DependsOn": Array [
        "CustomContentfulWebhook91f2075fb9504743a66bee0f6febf50dServiceRole311F9EBA",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters69f1a6347d9ec8846b5238785f2600d4d4498066a39df6942b0b4b7416efa262S3Bucket18224C47",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters69f1a6347d9ec8846b5238785f2600d4d4498066a39df6942b0b4b7416efa262S3VersionKeyED23AA03",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters69f1a6347d9ec8846b5238785f2600d4d4498066a39df6942b0b4b7416efa262S3VersionKeyED23AA03",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "lib/contentful-webhook.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomContentfulWebhook91f2075fb9504743a66bee0f6febf50dServiceRole311F9EBA",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomContentfulWebhook91f2075fb9504743a66bee0f6febf50dServiceRole311F9EBA": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
